src/layers/ordered_leaf.rs

```diff
@@ -0,0 +1,397 @@
+
+impl<K: Eq + Ord + Clone, R: Eq + Clone> Cursor for OrderedLeafCursor<K, R> {
+    type Key = (K, R);
+    type ChildKey = ();
+    type Storage = OrderedLeaf<K, R>;
+    type ValueStorage = ();
+
+    fn keys(&self) -> usize {
+        self.bounds.1 - self.bounds.0
+    }
+    fn key<'a>(&self, storage: &'a OrderedLeaf<K, R>) -> &'a Self::Key {
+        &storage.vals[self.pos]
+    }
+    fn values<'a>(&self, _storage: &'a OrderedLeaf<K, R>) -> (&'a (), ()) {
+        (&(), ())
+    }
+    fn step(&mut self, storage: &OrderedLeaf<K, R>) {
+        self.pos += 1;
+        if !self.valid(storage) {
+            self.pos = self.bounds.1;
+        }
+    }
+    fn seek(&mut self, storage: &OrderedLeaf<K, R>, key: &Self::Key) {
+        self.pos += advance(&storage.vals[self.pos..self.bounds.1], |(k, _)| {
+            k.lt(&key.0)
+        });
+    }
+    fn valid(&self, _storage: &OrderedLeaf<K, R>) -> bool {
+        self.pos < self.bounds.1
+    }
+    fn rewind(&mut self, _storage: &OrderedLeaf<K, R>) {
+        self.pos = self.bounds.0;
+    }
+    fn reposition(&mut self, _storage: &OrderedLeaf<K, R>, lower: usize, upper: usize) {
+        self.pos = lower;
+        self.bounds = (lower, upper);
+    }
+}
+
+/// Sorts and consolidates a slice, returning the valid prefix length.
+pub fn consolidate_slice<T: Ord, R: HasZero + AddAssignByRef + Clone>(
+    slice: &mut [(T, R)],
+) -> usize {
+    // We could do an insertion-sort like initial scan which builds up sorted,
+    // consolidated runs. In a world where there are not many results, we may
+    // never even need to call in to merge sort.
+    slice.sort_unstable_by(|x, y| x.0.cmp(&y.0));
+
+    // Counts the number of distinct known-non-zero accumulations. Indexes the write
+    // location.
+    let mut offset = 0;
+    for index in 1..slice.len() {
+        // The following unsafe block elides various bounds checks, using the reasoning
+        // that `offset` is always strictly less than `index` at the beginning
+        // of each iteration. This is initially true, and in each iteration
+        // `offset` can increase by at most one (whereas `index` always
+        // increases by one). As `index` is always in bounds, and `offset` starts at
+        // zero, it too is always in bounds.
+        //
+        // LLVM appears to struggle to optimize out Rust's split_at_mut, which would
+        // prove disjointness using run-time tests.
+        unsafe {
+            assert!(offset < index);
+
+            // LOOP INVARIANT: offset < index
+            let ptr1 = slice.as_mut_ptr().add(offset);
+            let ptr2 = slice.as_mut_ptr().add(index);
+
+            if (*ptr1).0 == (*ptr2).0 {
+                (*ptr1).1.add_assign_by_ref(&(*ptr2).1);
+            } else {
+                if !(*ptr1).1.is_zero() {
+                    offset += 1;
+                }
+                let ptr1 = slice.as_mut_ptr().add(offset);
+                std::ptr::swap(ptr1, ptr2);
+            }
```

Comment:
I'd make explicit that this is a different value from the initial `ptr1` since at a glance the two look identical (and like you could remove the second `let ptr = ...` until you recognize the `offset += 1` above it
```suggestion
            let mut ptr1 = slice.as_mut_ptr().add(offset);
            let ptr2 = slice.as_mut_ptr().add(index);

            if (*ptr1).0 == (*ptr2).0 {
                (*ptr1).1.add_assign_by_ref(&(*ptr2).1);
            } else {
                if !(*ptr1).1.is_zero() {
                    offset += 1;
                    ptr1 = ptr1.add(1);
                }

                std::ptr::swap(ptr1, ptr2);
            }
```